### PR TITLE
Social Preview: Update Facebook Preview

### PR DESCRIPTION
**Pull request by monsieur-z**
Related to 1203820933396971-as-1204281808721008

## Proposed Changes

This PR updates the `FacebookPreview` component of the `social-previews` package to match the existing Facebook site and post previews. It also updates the components instantiating `FacebookPreview` accordingly. Further changes will be made in the [jetpack monorepo](https://github.com/automattic/jetpack) to update the preview in WPadmin.

## Implementation Notes

- The render of the various previews was determined based on the actual renders on Facebook (see captures below).
- Facebook seems to truncate texts so that they don't take a certain amount of lines (or so that some elements don't exceed a certain height). This implementation uses a maximum number of characters instead. It's a good approximation for now, but that could be improved in a subsequent PR.
- Links in the custom message are not supported yet. It will be handled in a separate PR.
- Tests will be completed in a later PR.

### Prerequisites
- Make sure you have a Facebook account.
- Make sure you have a self-hosted site connected to Jetpack, with a subscription that includes Social.
- From the _Social_ or _My Jetpack_ page of your site WPadmin, activate Social.
- Then, connect your site to Facebook (check [these instructions](https://jetpack.com/support/jetpack-social/connecting-to-social-networks/) if you need help). You may need to create a test page.
- Spin up WordPress.com, by using a live link below or by running this branch locally.

### Testing a site

- Visit `marketing/traffic/:site`
- In the _Website Meta_ section, click _Show Previews_
- Select the Facebook Preview
- The preview of your site should look like this:
Screenshot 2023-04-11 at 10 04 09 AM

### Testing a post

- Create a new post from your site WPadmin
- In the Jetpack sidebar, make sure the post will be published automatically to your Facebook page
Screenshot 2023-04-11 at 4 43 17 PM

- Publish it
- Visit `/posts/:site` in Calypso
- You should see the newly created post under the _Published_ tab
- In the post menu, click _Share_, then press the _Preview_ button
Screenshot 2023-04-11 at 4 47 11 PM

- Select the Facebook preview
- Notice that it looks like the preview on Facebook, with the caveats mentioned in the **Implementation Notes** section
- Optionally, test different post variations as illustrated below

_Don't mind the fact that the user actions are not shown in the Facebook.com captures._
| Post | WordPress.com | Facebook.com |
|------|-----------------|----------------|
| Title only |Screenshot 2023-04-11 at 10 08 12 AM|Screenshot 2023-04-11 at 10 08 55 AM|
| Description only |Screenshot 2023-04-11 at 10 10 19 AM|Screenshot 2023-04-11 at 10 10 26 AM|
| Title & Description |Screenshot 2023-04-11 at 10 12 31 AM|Screenshot 2023-04-11 at 10 11 38 AM|
| Excerpt |Screenshot 2023-04-11 at 10 21 42 AM|Screenshot 2023-04-11 at 10 14 40 AM|
| Custom message |Screenshot 2023-04-11 at 11 00 50 AM|Screenshot 2023-04-11 at 11 00 38 AM|
| Landscape image |Screenshot 2023-04-11 at 3 37 49 PM|Screenshot 2023-04-11 at 3 37 19 PM|
| Portrait Image |Screenshot 2023-04-11 at 3 47 47 PM|Screenshot 2023-04-11 at 3 42 25 PM|

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204281808721008

### CODE REVIEW

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7920528

Thank you @monsieur-z for including a screenshot in the description! This is really helpful for our translators.

File: client/blocks/sharing-preview-pane/index.jsx [hunk 1]
```diff
@@ -111,7 +111,13 @@ class SharingPreviewPane extends PureComponent {
 
 		switch ( selectedService ) {
 			case 'facebook':
-				return <FacebookSharePreview { ...previewProps } />;
+				return (
+					<FacebookSharePreview
+						{ ...previewProps }
+						articleExcerpt={ post.excerpt }
```
Comment (monsieur-z) [inline client/blocks/sharing-preview-pane/index.jsx:117]: The Facebook preview handles truncation itself. We need to pass the raw excerpt and content to respect that.

File: client/components/share/facebook-share-preview/index.jsx [hunk 2]
```diff
@@ -1,119 +1,52 @@
...
-			externalProfileUrl,
 			imageUrl,
 			message,
 			seoTitle,
-			siteIcon,
-			translate,
 		} = this.props;
-		const { isProfileImageBroken } = this.state;
-		return (
-			<div className="facebook-share-preview">
```
Comment (monsieur-z) [inline client/components/share/facebook-share-preview/index.jsx:44]: This markup is not relevant anymore.

File: client/components/share/style.scss [hunk 3]
```diff
@@ -1,118 +1,4 @@
 //Mostly copied from Seo Preview
-.facebook-share-preview {
```
Comment (monsieur-z) [inline client/components/share/style.scss:2]: These styles are not relevant anymore.

File: packages/social-previews/src/facebook-preview/link-preview.tsx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	return (
+		<div className="facebook-preview__post">
+			<FacebookPostHeader user={ user } />
+			<div className="facebook-preview__content">
+				{ customText && (
+					<p className="facebook-preview__custom-text">{ facebookCustomText( customText ) }</p>
+				) }
+				<div className={ `facebook-preview__body ${ getModeClass() }` }>
+					{ ( image || isArticle ) && (
+						<div
```
Comment (monsieur-z) [inline packages/social-previews/src/facebook-preview/link-preview.tsx:42]: For blog posts, Facebook still displays a white square when there's no image.

File: packages/social-previews/src/facebook-preview/style.scss [hunk 5]
```diff
@@ -1,155 +1,137 @@
...
-	display: flex;
-	overflow-x: auto;
-	max-width: 527px;
-	margin: 20px;
-	-webkit-overflow-scrolling: touch;
+@import "./variables";
+@import "../shared";
+
+.facebook-preview__section {
+	width: clamp(#{$facebook-preview-min-width}, 100%, #{$facebook-preview-max-width});
```
Comment (monsieur-z) [inline packages/social-previews/src/facebook-preview/style.scss:5]: Facebook cards have a max-width of 680px, and a min-width of 508px.

File: packages/social-previews/src/facebook-preview/link-preview.tsx [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+	type,
+} ) => {
+	const [ mode, setMode ] = useState( LANDSCAPE_MODE );
+	const isArticle = type === TYPE_ARTICLE;
+	const portraitMode = ( isArticle && ! image ) || mode === PORTRAIT_MODE;
+
+	const getModeClass = useCallback(
+		() => `is-${ portraitMode ? 'portrait' : 'landscape' }`,
+		[ portraitMode ]
+	);
```
Comment (manzoorwanijk) [inline packages/social-previews/src/facebook-preview/link-preview.tsx]: `useCallback` is an overkill for this. a direct calculation will be much faster than `useCallback`.

```suggestion
	const modeClass = `is-${ portraitMode ? 'portrait' : 'landscape' }`;
```

**Review by manzoorwanijk — APPROVED** [review]
This looks great 🚀

File: packages/social-previews/src/facebook-preview/link-preview.tsx (hunk 6 above)
Comment (monsieur-z) [reply packages/social-previews/src/facebook-preview/link-preview.tsx]: That's right!

File: packages/social-previews/src/facebook-preview/index.tsx [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+import './style.scss';
+
+const FacebookPreview: React.FC< FacebookPreviewProps > = ( props ) => {
+	return (
+		<div className="social-preview facebook-preview">
+			<section className="social-preview__section facebook-preview__section">
+				<h3 className="social-preview__section-heading">
+					{
+						// translators: refers to a social post on Facebook
+						__( 'Your post' )
```
Comment (danielpost) [inline packages/social-previews/src/facebook-preview/index.tsx]: Do these need a textdomain? I'm not sure how this is usually handled.

File: packages/social-previews/src/facebook-preview/post/actions/index.tsx [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+			</li>
+			<li>
+				<FacebookPostIcon name="share" />
+				<span>
+					{
+						// translators: Facebook "Share" action
+						__( 'Share' )
+					}
+				</span>
+			</li>
```
Comment (danielpost) [inline packages/social-previews/src/facebook-preview/post/actions/index.tsx]: For this component, would it be nicer to have an object of all possible actions and loop through it? Something like this (untested):

```jsx
const actions = [
  {
    icon: 'like',
    // translators: Facebook "Like" action
    label: __( 'Like' )
  }
];

actions.map(({ icon, label }) => (
  <li>
    <FacebookPostIcon name={ icon } />
    <span>{ label }</span>
  </li>
))
```

Feel free to ignore this as well, I think it's fine either way!

File: packages/social-previews/src/facebook-preview/post/header/index.tsx [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+						</span>
+						<FacebookPostIcon name="public" />
+					</div>
+				</div>
+			</div>
+			<ul className="facebook-preview__post-header-more">
+				<li></li>
+				<li></li>
+				<li></li>
+			</ul>
```
Comment (danielpost) [inline packages/social-previews/src/facebook-preview/post/header/index.tsx]: You could try doing this with CSS and getting rid of all the extra markup, something like this:

```css
div {
  width: 24px;
  height: 4px;
  background-image: radial-gradient(circle at 2px 2px, black 2px, transparent 0);
  background-size: 10px 20px;
}
```

You can change the `10px` value to determine how close the dots are to each other.

File: packages/social-previews/src/facebook-preview/style.scss [hunk 10]
```diff
@@ -1,155 +1,137 @@
...
 
-		&::after {
-			content: "";
-			display: block;
-			height: 100%;
+	&.is-landscape {
+		> img {
 			width: 100%;
-			background: #fff;
+			aspect-ratio: 1.91;
```
Comment (danielpost) [inline packages/social-previews/src/facebook-preview/style.scss:110]: This comes from `1200 / 630` right? Do you think it's worth adding a comment to explain?

**Review by danielpost — APPROVED** [review]
This looks great, appreciate the detailed PR description and inline PR comments! I didn't functionally test it but looked through the code and it all looks good—just left a few minor thoughts/suggestions that you can implement or ignore as you see fit 👍🏼

File: packages/social-previews/src/facebook-preview/index.tsx (hunk 7 above)
Comment (monsieur-z) [reply packages/social-previews/src/facebook-preview/index.tsx]: Right, I forgot about it! I think it'd be better indeed.

File: packages/social-previews/src/facebook-preview/post/actions/index.tsx (hunk 8 above)
Comment (monsieur-z) [reply packages/social-previews/src/facebook-preview/post/actions/index.tsx]: I don't mind doing that for brevity :)

File: packages/social-previews/src/facebook-preview/style.scss (hunk 10 above)
Comment (monsieur-z) [reply packages/social-previews/src/facebook-preview/style.scss:110]: That's the ratio I measured on the Facebook page. The card can go from 508px to 680px wide. The image follows the same ratio at any width. I'll add a comment for clarity 👍

File: packages/social-previews/src/facebook-preview/post/header/index.tsx (hunk 9 above)
Comment (monsieur-z) [reply packages/social-previews/src/facebook-preview/post/header/index.tsx]: Good thinking!

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7920528).

**Pull request merged**